### DISCUSSION TITLE
Why Arabic, Persian, Kurdish letters not joined (they should be)?

### DISCUSSION THREAD

**Discussion by shashankaf** [wezterm/wezterm#5423] [Q&A]
### What Operating System(s) are you running on?

macOS

### WezTerm version

WezTerm 20240203-110809-5046fc22

### Ask your question!

Some of my projects have right to left texts injected into the UI and I use WezTerm as it has given me the best Terminal experience for that. The only problem I have, while the Terminal with my configuration shows these texts right to left (most other terminals don't), the letters are not joined, as they should be. No choice of font, whether mono or not, helps, I've tried even Arabic and Persian fonts (like `Vazir Font Hack`). However, the issue remains the same
Here's my WezTerm configuration
```
local wezterm = require 'wezterm'
local config = wezterm.config_builder()

config.font = wezterm.font 'FiraCode Nerd Font Mono'
config.font_size = 20
config.enable_tab_bar = false
config.window_background_opacity = 0.95
config.macos_window_background_blur = 30
config.window_decorations = "INTEGRATED_BUTTONS | RESIZE"
config.window_padding = {
  left = 10,
  right = 20,
  top = 30,
  bottom = 10,
}
config.bidi_enabled = true
return config
```

The correct way to disply the text is like this 
```
    if (phone.length < 3)
      setErrors([...errors, "تکایە ژمارەی مۆبایل داخڵ بکە"]);
```

but WezTerm shows it like this:
![Screenshot 2024-05-15 at 5 12 53 AM](https://github.com/wez/wezterm/assets/167884239/b80cb58c-195d-43d9-ba41-d13a75318a9b)

Does anyone have any idea how to solve this issue? Thanks in advance

**Comment by wez** [comment]
I'm not literate in this script, so forgive me for saying that what I see in the browser looks "the same" to me as the screenshot you shared:

<img width="816" alt="image" src="https://github.com/wez/wezterm/assets/117777/68b34796-98de-4441-a351-6862af37a51e">

Can you share more details on precisely where you're observing the issue? Perhaps by annotating the wezterm screenshot?

**Comment by zackAJ** [comment]
```
setErrors([...errors, "تکایە ژمارەی مۆبایل داخڵ بکە"]);
```
this is how it should render, I can confirm that that it's broken even for me:
![image](https://github.com/user-attachments/assets/3b8cb9c1-a565-41c6-8df3-7737e12993ba)

**Comment by masriomarm** [comment]
![image](https://github.com/user-attachments/assets/b08bcb8e-f7aa-4e4a-8770-f5ed40314ab6)
@wez Here's a SS compares `WezTerm` output to `Gnome terminal app` output.
Font seems to be written from left to right  in `WezTerm` - Arabic is drawn from right to left  unlike English, but chars are correct, only drawn wrong.
Issue is also present on `Alacritty` and `Foot`.

**Reply by wez** [reply]
Perhaps you need to enable Bidi mode?
Bidi in terminals is complex!
This comment summarizes the current level of bidi awareness/support in wezterm:
https://github.com/wezterm/wezterm/issues/784#issuecomment-1025346930
